File: dlitscustom/report/dlits_tax_report/dlits_tax_report.js
// Copyright (c) 2024, DLITS and contributors
// For license information, please see license.txt

frappe.query_reports["DLITS Tax Report"] = {
    "filters": [
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 1
        },
        {
            "fieldname": "tax_type",
            "label": __("Tax Type"),
            "fieldtype": "Data"
        }
    ],
    
    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        
        if (column.fieldname == "net_tax_amount" && data) {
            if (flt(data.net_tax_amount) > 0) {
                value = "<span style='color:green; font-weight:bold'>" + value + "</span>";
            } else if (flt(data.net_tax_amount) < 0) {
                value = "<span style='color:red; font-weight:bold'>" + value + "</span>";
            }
        }
        
        if (column.fieldname == "sales_tax_amount" && data && flt(data.sales_tax_amount) > 0) {
            value = "<span style='color:green'>" + value + "</span>";
        }
        
        if (column.fieldname == "purchase_tax_amount" && data && flt(data.purchase_tax_amount) > 0) {
            value = "<span style='color:blue'>" + value + "</span>";
        }
        
        return value;
    },
    
    "onload": function(report) {
        // Add PDF Export button
        report.page.add_inner_button(__("Export PDF"), function() {
            let filters = report.get_values();
            let data = report.data || [];
            generate_pdf_report(data, filters, "DLITS Tax Report");
        });
        
        // Add Print button
        report.page.add_inner_button(__("Print"), function() {
            let filters = report.get_values();
            let data = report.data || [];
            generate_pdf_report(data, filters, "DLITS Tax Report");
        });
    }
};

function get_filter_string(filters) {
    let filter_array = [];
    
    if (filters.from_date) {
        filter_array.push("From: " + frappe.datetime.str_to_user(filters.from_date));
    }
    if (filters.to_date) {
        filter_array.push("To: " + frappe.datetime.str_to_user(filters.to_date));
    }
    if (filters.tax_type) {
        filter_array.push("Tax Type: " + filters.tax_type);
    }
    
    return filter_array.join(" | ");
}

function generate_pdf_report(data, filters, report_name) {
    let filter_string = get_filter_string(filters);
    
    let html_content = `
        <div style="text-align: center; margin-bottom: 30px;">
            <h2 style="margin: 0; color: #333;">${report_name}</h2>
            <p style="margin: 5px 0; color: #666;">${filter_string}</p>
            <p style="margin: 5px 0; color: #666;">Generated on: ${frappe.datetime.str_to_user(frappe.datetime.get_today())}</p>
        </div>
        
        <table style="width: 100%; border-collapse: collapse; font-size: 12px;">
            <thead>
                <tr style="background-color: #f8f9fa;">
                    <th style="border: 1px solid #ddd; padding: 8px; text-align: left;">Tax Type</th>
                    <th style="border: 1px solid #ddd; padding: 8px; text-align: center;">Rate (%)</th>
                    <th style="border: 1px solid #ddd; padding: 8px; text-align: right;">Sales Tax</th>
                    <th style="border: 1px solid #ddd; padding: 8px; text-align: right;">Purchase Tax</th>
                    <th style="border: 1px solid #ddd; padding: 8px; text-align: right;">Net Tax</th>
                    <th style="border: 1px solid #ddd; padding: 8px; text-align: right;">Sales Base</th>
                    <th style="border: 1px solid #ddd; padding: 8px; text-align: right;">Purchase Base</th>
                    <th style="border: 1px solid #ddd; padding: 8px; text-align: center;">Sales Inv.</th>
                    <th style="border: 1px solid #ddd; padding: 8px; text-align: center;">Purchase Inv.</th>
                </tr>
            </thead>
            <tbody>`;
    
    let total_sales_tax = 0, total_purchase_tax = 0, total_net_tax = 0;
    let total_sales_base = 0, total_purchase_base = 0;
    let total_sales_invoices = 0, total_purchase_invoices = 0;
    
    data.forEach(function(row) {
        total_sales_tax += flt(row.sales_tax_amount);
        total_purchase_tax += flt(row.purchase_tax_amount);
        total_net_tax += flt(row.net_tax_amount);
        total_sales_base += flt(row.sales_base_amount);
        total_purchase_base += flt(row.purchase_base_amount);
        total_sales_invoices += flt(row.sales_invoices_count);
        total_purchase_invoices += flt(row.purchase_invoices_count);
        
        let net_tax_color = '';
        if (flt(row.net_tax_amount) > 0) {
            net_tax_color = 'color: green; font-weight: bold;';
        } else if (flt(row.net_tax_amount) < 0) {
            net_tax_color = 'color: red; font-weight: bold;';
        }
        
        html_content += `
            <tr>
                <td style="border: 1px solid #ddd; padding: 6px;">${row.tax_type}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: center;">${flt(row.tax_rate).toFixed(1)}%</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right; color: green;">${format_currency(row.sales_tax_amount)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right; color: blue;">${format_currency(row.purchase_tax_amount)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right; ${net_tax_color}">${format_currency(row.net_tax_amount)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right;">${format_currency(row.sales_base_amount)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right;">${format_currency(row.purchase_base_amount)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: center;">${flt(row.sales_invoices_count)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: center;">${flt(row.purchase_invoices_count)}</td>
            </tr>`;
    });
    
    // Add totals row
    let total_net_color = '';
    if (total_net_tax > 0) {
        total_net_color = 'color: green; font-weight: bold;';
    } else if (total_net_tax < 0) {
        total_net_color = 'color: red; font-weight: bold;';
    }
    
    html_content += `
            <tr style="background-color: #f8f9fa; font-weight: bold;">
                <td style="border: 1px solid #ddd; padding: 6px;" colspan="2">TOTAL</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right; color: green;">${format_currency(total_sales_tax)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right; color: blue;">${format_currency(total_purchase_tax)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right; ${total_net_color}">${format_currency(total_net_tax)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right;">${format_currency(total_sales_base)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: right;">${format_currency(total_purchase_base)}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: center;">${total_sales_invoices}</td>
                <td style="border: 1px solid #ddd; padding: 6px; text-align: center;">${total_purchase_invoices}</td>
            </tr>
        </tbody>
    </table>
    
    <div style="margin-top: 30px; font-size: 11px; color: #666;">
        <p><strong>Tax Summary:</strong></p>
        <ul>
            <li>Total Tax Types: ${data.length}</li>
            <li>Net Tax Position: ${format_currency(total_net_tax)} ${total_net_tax > 0 ? '(Tax Payable)' : total_net_tax < 0 ? '(Tax Refundable)' : '(Balanced)'}</li>
            <li>Total Sales Tax Collected: ${format_currency(total_sales_tax)}</li>
            <li>Total Purchase Tax Paid: ${format_currency(total_purchase_tax)}</li>
        </ul>
        <p><strong>Legend:</strong></p>
        <ul>
            <li><span style="color: green;">Green</span> = Sales Tax (Tax Collected)</li>
            <li><span style="color: blue;">Blue</span> = Purchase Tax (Tax Paid)</li>
            <li><span style="color: green; font-weight: bold;">Bold Green</span> = Net Tax Payable</li>
            <li><span style="color: red; font-weight: bold;">Bold Red</span> = Net Tax Refundable</li>
        </ul>
    </div>`;
    
    // Create and download PDF
    let pdf_window = window.open('', '_blank');
    pdf_window.document.write(`
        <html>
            <head>
                <title>${report_name}</title>
                <style>
                    body { font-family: Arial, sans-serif; margin: 20px; }
                    @media print { body { margin: 0; } }
                </style>
            </head>
            <body>
                ${html_content}
                <script>
                    window.onload = function() {
                        window.print();
                    }
                </script>
            </body>
        </html>
    `);
    pdf_window.document.close();
}

function format_currency(value) {
    return frappe.format(flt(value), {fieldtype: 'Currency'});
}

File: dlitscustom/report/dlits_tax_report/dlits_tax_report.py
# Copyright (c) 2024, DLITS and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, getdate, formatdate
from datetime import datetime, timedelta

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    chart = get_chart_data(data, filters)
    
    return columns, data, None, chart

def get_columns():
    return [
        {
            "fieldname": "tax_type",
            "label": _("Tax Type"),
            "fieldtype": "Data",
            "width": 120
        },
        {
            "fieldname": "tax_rate",
            "label": _("Tax Rate (%)"),
            "fieldtype": "Percent",
            "width": 100
        },
        {
            "fieldname": "sales_tax_amount",
            "label": _("Sales Tax"),
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "fieldname": "purchase_tax_amount",
            "label": _("Purchase Tax"),
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "fieldname": "net_tax_amount",
            "label": _("Net Tax"),
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "fieldname": "sales_base_amount",
            "label": _("Sales Base Amount"),
            "fieldtype": "Currency",
            "width": 140
        },
        {
            "fieldname": "purchase_base_amount",
            "label": _("Purchase Base Amount"),
            "fieldtype": "Currency",
            "width": 140
        },
        {
            "fieldname": "sales_invoices_count",
            "label": _("Sales Invoices"),
            "fieldtype": "Int",
            "width": 100
        },
        {
            "fieldname": "purchase_invoices_count",
            "label": _("Purchase Invoices"),
            "fieldtype": "Int",
            "width": 120
        }
    ]

def get_data(filters):
    conditions = get_conditions(filters)
    
    # Get sales tax data
    sales_tax_query = """
        SELECT 
            st.account_head as tax_account,
            st.rate as tax_rate,
            SUM(st.tax_amount) as tax_amount,
            SUM(st.base_tax_amount) as base_tax_amount,
            COUNT(DISTINCT si.name) as invoice_count,
            'Sales' as tax_type_source
        FROM `tabSales Taxes and Charges` st
        INNER JOIN `tabSales Invoice` si ON st.parent = si.name
        WHERE si.docstatus = 1 {conditions}
        GROUP BY st.account_head, st.rate
    """.format(conditions=conditions.replace("pi.", "si."))
    
    sales_tax_data = frappe.db.sql(sales_tax_query, filters, as_dict=True)
    
    # Get purchase tax data
    purchase_tax_query = """
        SELECT 
            pt.account_head as tax_account,
            pt.rate as tax_rate,
            SUM(pt.tax_amount) as tax_amount,
            SUM(pt.base_tax_amount) as base_tax_amount,
            COUNT(DISTINCT pi.name) as invoice_count,
            'Purchase' as tax_type_source
        FROM `tabPurchase Taxes and Charges` pt
        INNER JOIN `tabPurchase Invoice` pi ON pt.parent = pi.name
        WHERE pi.docstatus = 1 {conditions}
        GROUP BY pt.account_head, pt.rate
    """.format(conditions=conditions)
    
    purchase_tax_data = frappe.db.sql(purchase_tax_query, filters, as_dict=True)
    
    # Combine and process data
    tax_summary = {}
    
    # Process sales tax data
    for row in sales_tax_data:
        key = f"{row.tax_account}_{row.tax_rate}"
        if key not in tax_summary:
            tax_summary[key] = {
                "tax_type": get_tax_account_name(row.tax_account),
                "tax_rate": flt(row.tax_rate),
                "sales_tax_amount": 0,
                "purchase_tax_amount": 0,
                "sales_base_amount": 0,
                "purchase_base_amount": 0,
                "sales_invoices_count": 0,
                "purchase_invoices_count": 0
            }
        
        tax_summary[key]["sales_tax_amount"] += flt(row.base_tax_amount)
        tax_summary[key]["sales_base_amount"] += flt(row.base_tax_amount) / (flt(row.tax_rate) / 100) if flt(row.tax_rate) > 0 else 0
        tax_summary[key]["sales_invoices_count"] += flt(row.invoice_count)
    
    # Process purchase tax data
    for row in purchase_tax_data:
        key = f"{row.tax_account}_{row.tax_rate}"
        if key not in tax_summary:
            tax_summary[key] = {
                "tax_type": get_tax_account_name(row.tax_account),
                "tax_rate": flt(row.tax_rate),
                "sales_tax_amount": 0,
                "purchase_tax_amount": 0,
                "sales_base_amount": 0,
                "purchase_base_amount": 0,
                "sales_invoices_count": 0,
                "purchase_invoices_count": 0
            }
        
        tax_summary[key]["purchase_tax_amount"] += flt(row.base_tax_amount)
        tax_summary[key]["purchase_base_amount"] += flt(row.base_tax_amount) / (flt(row.tax_rate) / 100) if flt(row.tax_rate) > 0 else 0
        tax_summary[key]["purchase_invoices_count"] += flt(row.invoice_count)
    
    # Calculate net tax and convert to list
    data = []
    for key, values in tax_summary.items():
        values["net_tax_amount"] = flt(values["sales_tax_amount"]) - flt(values["purchase_tax_amount"])
        data.append(values)
    
    # Sort by tax rate and tax type
    data.sort(key=lambda x: (x["tax_rate"], x["tax_type"]))
    
    return data

def get_conditions(filters):
    conditions = ""
    
    if filters.get("from_date"):
        conditions += " AND pi.posting_date >= %(from_date)s"
    
    if filters.get("to_date"):
        conditions += " AND pi.posting_date <= %(to_date)s"
    
    if filters.get("tax_type"):
        conditions += " AND pt.account_head LIKE %(tax_type)s"
    
    return conditions

def get_tax_account_name(account_head):
    """Get a clean tax account name"""
    if not account_head:
        return "Unknown Tax"
    
    # Remove common prefixes and suffixes to get clean name
    clean_name = account_head.replace("VAT", "").replace("Tax", "").replace("Taxes", "")
    clean_name = clean_name.replace(" - ", "").strip()
    
    if not clean_name:
        return account_head
    
    return clean_name

def get_chart_data(data, filters):
    """Generate chart data for tax analysis"""
    if not data:
        return None
    
    # Tax distribution chart
    labels = []
    sales_values = []
    purchase_values = []
    
    for row in data:
        label = f"{row['tax_type']} ({row['tax_rate']}%)"
        labels.append(label)
        sales_values.append(flt(row["sales_tax_amount"]))
        purchase_values.append(flt(row["purchase_tax_amount"]))
    
    return {
        "data": {
            "labels": labels,
            "datasets": [
                {
                    "name": "Sales Tax",
                    "values": sales_values
                },
                {
                    "name": "Purchase Tax", 
                    "values": purchase_values
                }
            ]
        },
        "type": "bar",
        "height": 300,
        "colors": ["#28a745", "#dc3545"]
    }